pkg/util/span/frontier.go

```diff
@@ -185,229 +276,283 @@ func (f *Frontier) PeekFrontierSpan() roachpb.Span {
-	var toAdd, toRemove []*frontierEntry
+// clone augments generated iterStack code to support cloning.
+func (is *iterStack) clone() iterStack {
+	c := *is
+	c.s = append([]iterFrame(nil), is.s...) // copy stack.
+	return c
+}
+
+// clone augments generated iterator code to support cloning.
+func (i *iterator) clone() iterator {
+	c := *i
+	c.s = i.s.clone() // copy stack.
+	return c
+}
+
+// mergeEntries searches for the entries to the left and to the right
+// of the input entry that are contiguous to the entry range and have the same timestamp.
+// Updates btree to include single merged entry.
+// Any existing tree iterators are invalid after this call.
+// Returns btreeFrontierEntry that replaced passed in entry.
+func (f *btreeFrontier) mergeEntries(e *btreeFrontierEntry) (*btreeFrontierEntry, error) {
+	defer func() {
+		f.mergeAlloc = f.mergeAlloc[:0]
+	}()
+
+	// First, position iterator at e.
+	pos := f.tree.MakeIter()
+	pos.SeekGE(e)
+	if !pos.Valid() || pos.Cur() != e {
+		return nil, errors.AssertionFailedf("failed to find entry %s in btree", e)
+	}
 
-	// addEntry adds frontier entry to the toAdd list.
-	addEntry := func(r interval.Range, ts hlc.Timestamp) {
-		sp := makeSpan(r)
-		toAdd = append(toAdd, &frontierEntry{
-			id:   f.idAlloc,
-			span: sp,
-			keys: sp.AsRange(),
-			ts:   ts,
-		})
-		f.idAlloc++
+	// Now, search for contiguous spans to the left of e.
+	leftMost := e
+	leftIter := pos.clone()
+	for leftIter.Prev(); leftIter.Valid(); leftIter.Prev() {
+		if !(leftIter.Cur().End.Equal(leftMost.Start) && leftIter.Cur().ts.Equal(e.ts)) {
+			break
+		}
+		f.mergeAlloc = append(f.mergeAlloc, leftIter.Cur())
+		leftMost = leftIter.Cur()
 	}
 
-	// todoRange is the range we're adding. It gets updated as we process the range.
-	todoRange := span.AsRange()
+	end := leftMost.End
+	if leftMost != e {
+		// We found ranges to the left of e that have the same timestamp.
+		// That means that we'll merge entries into leftMost, and we will
+		// also subsume e itself.
+		f.mergeAlloc[len(f.mergeAlloc)-1] = e
+		end = e.End
+	}
 
-	// pendingSpan (if not empty) is the span of multiple overlap intervals
-	// we'll merge together (because all of those intervals have timestamp lower
-	// than insertTS).
-	var pendingSpan interval.Range
+	// Now, continue to the right of e.
+	rightIter := pos.clone()
+	for rightIter.Next(); rightIter.Valid(); rightIter.Next() {
+		if !(rightIter.Cur().Start.Equal(end) && rightIter.Cur().ts.Equal(e.ts)) {
+			break
+		}
+		end = rightIter.Cur().End
+		f.mergeAlloc = append(f.mergeAlloc, rightIter.Cur())
+	}
 
-	// consumePrefix consumes todoRange prefix ending at 'end' and moves
-	// that prefix into pendingSpan.
-	consumePrefix := func(end interval.Comparable) {
-		if pendingSpan.Start == nil {
-			pendingSpan.Start = todoRange.Start
+	// Delete entries first, before updating leftMost boundaries since doing so
+	// will mess up btree.
+	for i, toRemove := range f.mergeAlloc {
```

Comment:
maybe I missed something, but didn't we add `lefMost` to the `mergeAlloc` here:
```
f.mergeAlloc = append(f.mergeAlloc, leftIter.Cur())
		leftMost = leftIter.Cur()
```
which means we will delete it from the frontier?

It seems that after this function returns, leftMost should be in the frontier though.